### PR TITLE
Improve grouping performance via  better vectorization in accumulate functions

### PR DESCRIPTION
**Pull request by alamb**
Draft as I am running performance tests and will clean up the code a bit if possible. 

# Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/6889

This is a follow on to https://github.com/apache/arrow-datafusion/pull/6904 where some last minute changes (to track null state correctly) decreased performance slightly -- I am trying to get some back

# Rationale for this change
I am trying to help the compiler to vectorize the inner loops of the accumulators. Let's do this by doing the minimal work as possible

# What changes are included in this PR?
Update null state with a separate loop

# Are these changes tested?
covered by existing tests

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/aggregate/groups_accumulator/accumulate.rs [hunk 1]
```diff
@@ -139,10 +139,14 @@ impl NullState {
             // no nulls, no filter,
             (false, None) => {
                 let iter = group_indices.iter().zip(data.iter());
+
                 for (&group_index, &new_value) in iter {
-                    seen_values.set_bit(group_index, true);
                     value_fn(group_index, new_value);
                 }
+                // update seen values in separate loop
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/groups_accumulator/accumulate.rs:146]: the idea is to use a separate loop

Note we can skip this update entirely once all the groups have been seen -- I will explore if I can figure out how to do that next.

**Comment by alamb** [comment]
This shows a little benefit bit maybe in the noise range

```
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃ main_base ┃ alamb_faster_grouping ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │  560.29ms │              533.10ms │     no change │
│ QQuery 2     │  174.72ms │              162.55ms │ +1.07x faster │
│ QQuery 3     │  167.44ms │              161.30ms │     no change │
│ QQuery 4     │  119.56ms │              117.52ms │     no change │
│ QQuery 5     │  394.33ms │              400.22ms │     no change │
│ QQuery 6     │   45.54ms │               40.08ms │ +1.14x faster │
│ QQuery 7     │  849.06ms │              837.12ms │     no change │
│ QQuery 8     │  258.03ms │              247.16ms │     no change │
│ QQuery 9     │  595.82ms │              566.07ms │     no change │
│ QQuery 10    │  341.31ms │              337.04ms │     no change │
│ QQuery 11    │  170.28ms │              159.25ms │ +1.07x faster │
│ QQuery 12    │  175.19ms │              173.80ms │     no change │
│ QQuery 13    │  331.46ms │              315.21ms │     no change │
│ QQuery 14    │   48.67ms │               49.25ms │     no change │
│ QQuery 15    │   59.04ms │               50.88ms │ +1.16x faster │
│ QQuery 16    │  164.45ms │              164.20ms │     no change │
│ QQuery 17    │  966.84ms │              873.41ms │ +1.11x faster │
│ QQuery 18    │ 1676.43ms │             1637.90ms │     no change │
│ QQuery 19    │  166.95ms │              167.24ms │     no change │
│ QQuery 20    │  313.92ms │              313.73ms │     no change │
│ QQuery 21    │ 1062.69ms │             1060.65ms │     no change │
│ QQuery 22    │   88.17ms │               93.33ms │  1.06x slower │
└──────────────┴───────────┴───────────────────────┴───────────────┘
```

**Comment by alamb** [comment]
I am now benchmarking the special case of "don't track nulls they can't possibly be present

**Comment by alamb** [comment]
Special casing the nulls seems to have helped a little, though the fact that some queries now report slower is surprising to me

```
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃ main_base ┃ alamb_faster_grouping ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │  574.53ms │              522.95ms │ +1.10x faster │
│ QQuery 2     │  168.33ms │              152.40ms │ +1.10x faster │
│ QQuery 3     │  173.45ms │              182.54ms │  1.05x slower │
│ QQuery 4     │  116.60ms │              116.16ms │     no change │
│ QQuery 5     │  402.46ms │              405.81ms │     no change │
│ QQuery 6     │   45.04ms │               41.94ms │ +1.07x faster │
│ QQuery 7     │  922.93ms │              916.25ms │     no change │
│ QQuery 8     │  255.73ms │              240.93ms │ +1.06x faster │
│ QQuery 9     │  577.15ms │              554.53ms │     no change │
│ QQuery 10    │  318.40ms │              330.10ms │     no change │
│ QQuery 11    │  167.56ms │              172.50ms │     no change │
│ QQuery 12    │  174.26ms │              177.85ms │     no change │
│ QQuery 13    │  324.26ms │              320.80ms │     no change │
│ QQuery 14    │   48.11ms │               48.24ms │     no change │
│ QQuery 15    │   54.18ms │               57.05ms │  1.05x slower │
│ QQuery 16    │  164.83ms │              168.32ms │     no change │
│ QQuery 17    │  897.29ms │              877.01ms │     no change │
│ QQuery 18    │ 1672.32ms │             1675.61ms │     no change │
│ QQuery 19    │  179.13ms │              166.64ms │ +1.07x faster │
│ QQuery 20    │  324.28ms │              324.27ms │     no change │
│ QQuery 21    │ 1147.35ms │             1177.47ms │     no change │
│ QQuery 22    │   93.86ms │               91.11ms │     no change │
```

**Comment by Dandandan** [comment]
Noise maybe? Your previous run seems to have some noisy results as well.
I only get relatively relatively stable perf results by closing all browsers / IDEs etc.

**Comment by alamb** [comment]
> Noise maybe? Your previous run seems to have some noisy results as well.
> I only get relatively relatively stable perf results by closing all browsers / IDEs etc.

Yeah, I am trying to control for noise by running these experiments on a GCP machine (not my laptop) 🤔

**Comment by alamb** [comment]
Tracking this basic idea in https://github.com/apache/arrow-datafusion/issues/7066 , I don't plan ti pursue this PR anymore for now

**Pull request closed without merging**